pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -0,0 +1,223 @@
+            metrics_client.histogram('pubsub.consumer.latency.runtime',
+                                     time.perf_counter() - start)
+        except Exception:
+            log.exception('Application callback raised an exception')
+            metrics_client.increment('pubsub.consumer.failed')
+
+    async def consumer(
+            message_queue: MessageQueue,
+            callback: ApplicationHandler,
+            ack_queue: 'asyncio.Queue[str]',
+            ack_deadline_cache: AckDeadlineCache,
+            consumer_pool_size: int,
+            metrics_client: MetricsAgent) -> None:
+        semaphore = asyncio.Semaphore(consumer_pool_size)
+        while True:
+            await semaphore.acquire()
+
+            message, pulled_at = await message_queue.get()
+
+            ack_deadline = await ack_deadline_cache.get()
+            if (time.perf_counter() - pulled_at) >= ack_deadline:
+                metrics_client.increment('pubsub.consumer.failfast')
+                message_queue.task_done()
+                semaphore.release()
+                continue
+
+            metrics_client.histogram(
+                'pubsub.consumer.latency.receive',
+                # publish_time is in UTC Zulu
+                # https://cloud.google.com/pubsub/docs/reference/rest/v1/PubsubMessage
+                time.time() - message.publish_time.timestamp())
+
+            task = asyncio.ensure_future(_execute_callback(
+                message,
+                callback,
+                ack_queue,
+                metrics_client,
+            ))
+            task.add_done_callback(lambda _f: semaphore.release())
+            message_queue.task_done()
+
+    async def producer(
+            subscription: str,
+            message_queue: MessageQueue,
+            subscriber_client: 'SubscriberClient',
+            max_messages: int,
+            metrics_client: MetricsAgent
+    ) -> None:
+        while True:
+            try:
+                new_messages = await subscriber_client.pull(
+                    subscription=subscription, max_messages=max_messages)
+            except asyncio.TimeoutError:
+                continue
+
+            pulled_at = time.perf_counter()
+            for m in new_messages:
+                await message_queue.put((m, pulled_at))
+
+            metrics_client.histogram(
+                'pubsub.producer.batch', len(new_messages))
+
+            await message_queue.join()
+
+    async def subscribe(subscription: str,  # pylint: disable=too-many-locals
+                        handler: ApplicationHandler,
+                        subscriber_client: SubscriberClient,
+                        *,
+                        num_workers: int = 1,
+                        max_messages: int = 100,
+                        ack_window: float = 0.3,
+                        ack_deadline_cache_timeout: int = 10,
+                        consumer_pool_size: int = 1,
```

Comment:
This might be better named `num_consumers_per_worker` to line it up with the terminology used in `num_workers` and proposed `max_messages_per_worker`.

Relatedly, `worker` is a bit broad with all the various types of workers we have here (eg. it might not be obvious that worker here means "pair of producers and n-thread0consumers". Perhaps having `num_producers`, `max_messages_per_producer`, and `num_consumers_per_producer` would be nice and clear?

Reply:
alright so now when the design is settled time to decide on this. We can go with "imaginary composite workers" and name parameters `num_workers`, `max_messages_per_worker` and `worker_pool_size`. Or we may not try to hide implementation complexity from users and call them simply "num_producers", "max_messages_per_producer", "max_tasks_per_consumer"? I'm leaning towards the latter one, since we'll probably have to expose implementation in the docs anyway. And of course feel free to nominate your suggestions :)

/cc @TheKevJames @shaundialpad @jonathan-johnston @allandialpad

Reply:
I like hiding implementation details, but I like clarity even more I think. Without any context, the former option would be very confusing to me and I'd have to dive into the code to see what each setting is referring to, particularly the difference between `num_workers` and `worker_pool_size`. That being said, even with the latter option, I would not know exactly how I should go about tuning these anyhow. I suppose we're going to have some sane defaults and then elaborate on how one might do this in the documentation anyway.

So yeah I vote for `num_producers`, `max_messages_per_producer` and `num_consumers_per_producer`

Reply:
Agreed with option 2 :+1: `num_tasks_per_consumer` makes sense for now, and we can change it to `num_consumers_per_producer` if we ever figure out what's up with the perf on that and switch over.